File: src/dexbox/session.py
"""Session manager — tracks active workflow sessions"""

from __future__ import annotations

import asyncio
import logging
import secrets
import time
from dataclasses import dataclass, field
from typing import Any

logger = logging.getLogger("dexbox.session")

SESSION_TIMEOUT_SECONDS = 3600  # 1 hour


@dataclass
class WorkflowSession:
    """State for a single workflow execution."""

    token: str
    api_key: str
    model: str
    provider: str = "anthropic"
    anthropic_base_url: str | None = None
    workflow_id: str = ""
    variables: dict[str, Any] = field(default_factory=dict)
    secure_params: dict[str, str] = field(default_factory=dict)
    artifacts_dir: str | None = None
    created_at: float = field(default_factory=time.time)
    terminated: bool = False

    # Set by orchestrator during execution
    input_data: dict[str, Any] | None = None
    code: str | None = None
    output_data: dict[str, Any] | None = None
    assets_archive: bytes | None = None

    # Tool collection is set lazily by internal_routes
    _tool_collection: Any = field(default=None, repr=False)

    # Active async tasks for cancellation
    _active_tasks: set[asyncio.Task] = field(default_factory=set, repr=False)

    # Event queue for real-time progress streaming
    event_queue: asyncio.Queue = field(default_factory=asyncio.Queue, repr=False)

    def is_expired(self) -> bool:
        return time.time() - self.created_at > SESSION_TIMEOUT_SECONDS

    def register_task(self, task: asyncio.Task) -> None:
        """Register an async task for lifecycle management."""
        self._active_tasks.add(task)
        task.add_done_callback(self._active_tasks.discard)

    def cancel_all_tasks(self) -> int:
        """Cancel all active tasks. Returns number cancelled."""
        tasks = list(self._active_tasks)
        for task in tasks:
            task.cancel()
        return len(tasks)


class SessionManager:
    """Manages active workflow sessions."""

    def __init__(self) -> None:
        self._sessions: dict[str, WorkflowSession] = {}

    @property
    def active_session_count(self) -> int:
        return sum(1 for s in self._sessions.values() if not s.is_expired() and not s.terminated)

    def create_session(
        self,
        *,
        api_key: str,
        model: str,
        provider: str = "anthropic",
        anthropic_base_url: str | None = None,
        workflow_id: str = "",
        variables: dict[str, Any] | None = None,
        secure_params: dict[str, str] | None = None,
        artifacts_dir: str | None = None,
    ) -> str:
        """Create a session and return its token."""
        token = secrets.token_urlsafe(32)
        session = WorkflowSession(
            token=token,
            api_key=api_key,
            model=model,
            provider=provider,
            anthropic_base_url=anthropic_base_url,
            workflow_id=workflow_id,
            variables=variables or {},
            secure_params=secure_params or {},
            artifacts_dir=artifacts_dir,
        )
        self._sessions[token] = session
        logger.info("Created session %s for workflow %s", token[:8], workflow_id)
        self._cleanup_expired()
        return token

    def get_session(self, token: str) -> WorkflowSession | None:
        """Return session if valid (not expired/terminated), else None."""
        session = self._sessions.get(token)
        if session is None:
            return None
        if session.terminated:
            return None
        if session.is_expired():
            del self._sessions[token]
            return None
        return session

    def get_session_status(self, token: str) -> tuple[WorkflowSession | None, str | None]:
        """Return (session, error) where error is 'not_found'|'terminated'|'expired'|None."""
        session = self._sessions.get(token)
        if session is None:
            return None, "not_found"
        if session.terminated:
            return None, "terminated"
        if session.is_expired():
            return None, "expired"
        return session, None

    def delete_session(self, token: str) -> bool:
        """Delete session. Returns True if it existed."""
        if token in self._sessions:
            del self._sessions[token]
            return True
        return False

    def terminate_session(self, token: str) -> bool:
        """Mark session as terminated and cancel all tasks."""
        session = self._sessions.get(token)
        if session is None:
            return False
        session.terminated = True
        session.cancel_all_tasks()
        logger.info("Terminated session %s", token[:8])
        return True

    def _cleanup_expired(self) -> None:
        expired = [t for t, s in self._sessions.items() if s.is_expired()]
        for token in expired:
            del self._sessions[token]
        if expired:
            logger.debug("Cleaned up %d expired sessions", len(expired))

File: internal/server/session.go
package server

import (
	"fmt"
	"sync"
	"time"

	"github.com/google/uuid"
)

type Session struct {
	Token            string              `json:"token"`
	APIKey           string              `json:"api_key"`
	AnthropicAPIKey  string              `json:"anthropic_api_key,omitempty"`
	OpenAIAPIKey     string              `json:"openai_api_key,omitempty"`
	LuxAPIKey        string              `json:"lux_api_key,omitempty"`
	GeminiAPIKey     string              `json:"gemini_api_key,omitempty"`
	Model            string              `json:"model"`
	Provider         string              `json:"provider"`
	AnthropicBaseURL string              `json:"anthropic_base_url,omitempty"`
	OpenAIBaseURL    string              `json:"openai_base_url,omitempty"`
	WorkflowID       string              `json:"workflow_id"`
	Variables        map[string]any      `json:"variables,omitempty"`
	SecureParams     map[string]string   `json:"secure_params,omitempty"`
	ArtifactsDir     string              `json:"artifacts_dir,omitempty"`
	CreatedAt        time.Time           `json:"created_at"`
	Code             string              `json:"code,omitempty"`
	EventQueue       chan map[string]any `json:"-"`
	OutputData       map[string]any      `json:"output,omitempty"`
}

type SessionManager struct {
	mu       sync.RWMutex
	sessions map[string]*Session
}

func NewSessionManager() *SessionManager {
	return &SessionManager{
		sessions: make(map[string]*Session),
	}
}

func (m *SessionManager) CreateSession(s *Session) string {
	m.mu.Lock()
	defer m.mu.Unlock()

	if s.Token == "" {
		s.Token = uuid.New().String()
	}
	s.CreatedAt = time.Now()
	s.EventQueue = make(chan map[string]any, 100)
	m.sessions[s.Token] = s
	return s.Token
}

func (m *SessionManager) GetSession(token string) (*Session, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	s, ok := m.sessions[token]
	if !ok {
		return nil, fmt.Errorf("session not found")
	}
	return s, nil
}

func (m *SessionManager) DeleteSession(token string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.sessions, token)
}
